### PR TITLE
Enabling Serverside Rendering Support

### PR DESCRIPTION
**Pull request by Schachte**
Enables the ability to opt-into server-side rendering performance improvements for compatible applications. 

`yarn start --enable-server-rendering`

### CODE REVIEW

File: lib/start.js [hunk 1]
```diff
@@ -29,23 +31,46 @@ module.exports = ({ args, getPackage }) => {
...
+    devServerConfiguration
+  )
+
+  // Server metadata for routes/middleware
+  let clientBundle = null
+  clientCompiler.hooks.emit.tap('CompiledFilePlugin', compilation => {
+    // clientBundle = Object.keys(compilation.assets).filter(
+    //   name => name.match(/\.js$/)[0]
+    // )
+    clientBundle = Object.keys(compilation.assets)[0]
```
Comment (Schachte) [inline lib/start.js]: Need to filter on regex for a valid bundle file

File: lib/webpack.config.js [hunk 2]
```diff
@@ -75,6 +76,34 @@ const generateExcludeFromWebpackBuild = packageJson => {
...
+const server = base =>
+  merge.smart(base, {
+    target: 'node',
+    mode: 'development',
+    devtool: 'cheap-module-eval-source-map',
+
+    // Tell webpack the root file of our
+    // server application
+    entry:
+      '/Users/schachte/Desktop/connextaUi/revelio/src/main/graphql-server/helpers/renderer.js',
```
Comment (Schachte) [inline lib/webpack.config.js]: Needs to be configurable

File: lib/initializeMiddleware.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      renderer = requireFromString(
+        contents,
+        serverWebpackConfig.output.filename
+      ).default
+      resolve()
+    })
+  })
+
+  // Upstream or (((downstream))) for this definition?
+  const ssrPaths = [
```
Comment (djblue) [inline lib/initializeMiddleware.js]: This should be downstream

File: lib/initializeMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+const webpack = require('webpack')
+const path = require('path')
+
+require('source-map-support').install({
+  environment: 'node',
+  hookRequire: true,
+})
+
+module.exports = (app, serverWebpackConfig) => {
+  let renderer = null
```
Comment (djblue) [inline lib/initializeMiddleware.js]: I think this is more of middleware now

File: lib/initializeMiddleware.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  const ready = new Promise((resolve, reject) => {
+    const fs = new MemoryFS()
+    const serverCompiler = webpack(serverWebpackConfig)
+
+    // Ensure we write the contents of the bundle to memory instead of to disk
+    serverCompiler.outputFileSystem = fs
+
+    // Run webpack to generate the bundle and write into memory
+    serverCompiler.watch({}, (err, stats) => {
+      console.log('Watch here')
```
Comment (djblue) [inline lib/initializeMiddleware.js]: remove

File: lib/initializeMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      if (err) {
+        console.error(err)
+        return err
```
Comment (djblue) [inline lib/initializeMiddleware.js]: We could reject the promise here which would signal an initialization failure

Comment (djblue) [reply lib/initializeMiddleware.js]: `return reject(err)`

File: lib/initializeMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
+const requireFromString = require('require-from-string')
```
Comment (djblue) [inline lib/initializeMiddleware.js]: I would go with `initialize-middleware.js ` to match all the other files

File: lib/start.js [hunk 8]
```diff
@@ -2,20 +2,22 @@ const openBrowser = require('react-dev-utils/openBrowser')
...
 
 const webpackConfig = require('./webpack.config')
 
 module.exports = ({ args, getPackage }) => {
+  const app = express()
   const pkg = getPackage()
-  const publicPath = args.contextPath || pkg['context-path']
+  const publicPath = pkg['context-path']
 
-  const main = args.main || pkg.main
```
Comment (djblue) [inline lib/start.js]: Undo, I think you might have applied changes against an older version of ace. This is a breaking change.

File: lib/start.js [hunk 9]
```diff
@@ -2,20 +2,22 @@ const openBrowser = require('react-dev-utils/openBrowser')
...
 const chalk = require('chalk')
+const initializeMiddleware = require('./initializeMiddleware')
+const express = require('express')
 
 const webpackConfig = require('./webpack.config')
 
 module.exports = ({ args, getPackage }) => {
+  const app = express()
   const pkg = getPackage()
-  const publicPath = args.contextPath || pkg['context-path']
```
Comment (djblue) [inline lib/start.js]: Undo

File: lib/start.js [hunk 10]
```diff
@@ -2,20 +2,22 @@ const openBrowser = require('react-dev-utils/openBrowser')
...
+  const publicPath = pkg['context-path']
 
-  const main = args.main || pkg.main
   const env = process.env.NODE_ENV || args.env || 'development'
   const port = process.env.PORT || args.port || 8080
   const host = process.env.HOST || args.host || 'localhost'
   const proxy = process.env.PROXY || args.proxy || 'https://localhost:8993'
 
-  const config = webpackConfig({
+  const { client, server } = webpackConfig({
```
Comment (djblue) [inline lib/start.js]: I think we should change this to return the server or the client config, not both.

File: lib/start.js [hunk 11]
```diff
@@ -29,23 +31,46 @@ module.exports = ({ args, getPackage }) => {
       'admin:admin',
 
     tsTranspileOnly: args.tsTranspileOnly === 'true',
-    main,
+    main: pkg.main,
```
Comment (djblue) [inline lib/start.js]: Undo

File: lib/start.js [hunk 12]
```diff
@@ -29,23 +31,46 @@ module.exports = ({ args, getPackage }) => {
...
+  let clientBundle = null
+  clientCompiler.hooks.emit.tap('CompiledFilePlugin', compilation => {
+    // clientBundle = Object.keys(compilation.assets).filter(
+    //   name => name.match(/\.js$/)[0]
+    // )
+    clientBundle = Object.keys(compilation.assets)[0]
+  })
+  app.use((req, res, next) => {
+    req.clientBundle = clientBundle
+    console.log(req.clientBundle + ' from the middleware')
```
Comment (djblue) [inline lib/start.js]: remove

File: lib/start.js [hunk 13]
```diff
@@ -29,23 +31,46 @@ module.exports = ({ args, getPackage }) => {
...
+  // Utilize composition to extend the webpackdevserver routes onto our custom server
+  app.use(webpackDevServer.app)
+
+  app.listen(port, host, err => {
     if (err) {
       console.error(err)
       process.exit(1)
     } else if (args.open) {
-      openBrowser(`http://${host}:${port}${publicPath}/`)
+      // openBrowser(`http://${host}:${port}${publicPath}/`)
```
Comment (djblue) [inline lib/start.js]: undo

File: lib/webpack.config.js [hunk 14]
```diff
@@ -89,10 +118,6 @@ const base = ({ alias = {}, env, tsTranspileOnly, packageJson }) => ({
     //   analyzerMode: 'static',
     //   reportFilename: resolve('target/report.html'),
     // }),
-    new webpack.ContextReplacementPlugin(
```
Comment (djblue) [inline lib/webpack.config.js]: undo

File: lib/webpack.config.js [hunk 15]
```diff
@@ -344,12 +383,13 @@ const dev = (base, { main, packageJson }) => {
         filename: 'index.html',
         template: getTemplateFile(packageJson),
       }),
-      new SimpleProgressWebpackPlugin({
-        format: 'compact',
-      }),
+      // new SimpleProgressWebpackPlugin({
```
Comment (Schachte) [inline lib/webpack.config.js]: ❓ Should we just remove this completely now?

File: lib/webpack.config.js [hunk 16]
```diff
@@ -332,6 +362,15 @@ const getTemplateFile = packageJson => {
 }
 
 const dev = (base, { main, packageJson }) => {
+  let template,
```
Comment (Schachte) [inline lib/webpack.config.js]: ❓ Should we get rid of all this?

**Comment by Schachte** [comment]
Need to add support to conditionally write the serverside bundle to disk and not just the FS memory for production builds. (bundle.js)

File: lib/start.js [hunk 17]
```diff
@@ -32,15 +38,39 @@ module.exports = ({ args, getPackage }) => {
...
+      next()
+    })
+
+    initializeMiddleware(app, server)
+  }
 
-  WebpackDevServer.addDevServerEntrypoints(config, devServer)
-  const compiler = webpack(config)
-  const server = new WebpackDevServer(compiler, devServer)
+  app.use(webpackDevServer.app)
```
Comment (djblue) [inline lib/start.js]:
```javascript
webpackDevServer.app = app
webpackDevServer.listen(...)
```

Comment (djblue) [reply lib/start.js]: This will fix the hot reloading issue :ok_hand:

**Comment by Bdthomson** [comment]
What breaking changes (if any) to `package.json` does this require in a project that doens't plan on using server side rendering yet?

**Comment by Schachte** [comment]
> What breaking changes (if any) to `package.json` does this require in a project that doesn't plan on using server side rendering yet?

It's opt-in only feature that won't have breaking changes

File: package.json [hunk 18]
```diff
@@ -18,6 +18,7 @@
   },
   "dependencies": {
     "@babel/core": "^7.0.0",
+    "@babel/polyfill": "^7.7.0",
```
Comment (djblue) [inline package.json]: Can you remove this and revert the yarn.lock?

File: lib/storybook/webpack.config.js [hunk 19]
```diff
@@ -1,12 +1,10 @@
...
 const merge = require('webpack-merge')
-const webpackConfig = require('../webpack.config')
-const path = require('path')
+const { retrieveClientConfig } = require('./webpack.config')
 
 module.exports = ({ config, mode }) => {
   return merge.smart(
     config,
-    webpackConfig({
+    retrieveClientConfig({
```
Comment (djblue) [inline lib/storybook/webpack.config.js]: Good catch!

**Pull request merged**